proxy/GcsConnection.go

```diff
@@ -0,0 +1,300 @@
+package proxy
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+)
+
+func (rm *RequestMessage) Read(p []byte) (n int, err error) {
+	log.Println("RequestMessage: Starting Read")
+	headerBuf := make([]byte, binary.Size(rm.header))
+	err = binary.Write(io.Discard, binary.LittleEndian, rm.header)
+	if err != nil {
+		log.Printf("RequestMessage: Error writing header: %v", err)
+		return 0, err
+	}
+	copy(p, headerBuf)
+	n += len(headerBuf)
+
+	if rm.secondHeader != nil {
+		nSecond, err := rm.secondHeader.Read(p[n:])
+		n += nSecond
+		if err != nil && err != io.EOF {
+			log.Printf("RequestMessage: Error reading secondHeader: %v", err)
+			return n, err
+		}
+		if err == io.EOF && rm.data != nil {
+			nData, err := rm.data.Read(p[n:])
+			n += nData
+			return n, err
+		}
+	}
+	if rm.data != nil {
+		nData, err := rm.data.Read(p[n:])
+		n += nData
+		return n, err
+	}
+	log.Println("RequestMessage: Finished Read with EOF")
+	return n, io.EOF
+}
+
+func (resp *ResponseMessage) Read(p []byte) (n int, err error) {
+	log.Println("ResponseMessage: Starting Read")
+	headerBuf := make([]byte, binary.Size(resp.header))
+	err = binary.Write(io.Discard, binary.LittleEndian, resp.header)
+	if err != nil {
+		log.Printf("ResponseMessage: Error writing header: %v", err)
+		return 0, err
+	}
+	copy(p, headerBuf)
+	n += len(headerBuf)
+
+	if resp.secondHeader != nil {
+		nSecond, err := resp.secondHeader.Read(p[n:])
+		n += nSecond
+		if err != nil && err != io.EOF {
+			log.Printf("ResponseMessage: Error reading secondHeader: %v", err)
+			return n, err
+		}
+		if err == io.EOF && resp.data != nil {
+			nData, err := resp.data.Read(p[n:])
+			n += nData
+			return n, err
+		}
+	}
+	if resp.data != nil {
+		nData, err := resp.data.Read(p[n:])
+		n += nData
+		return n, err
+	}
+	log.Println("ResponseMessage: Finished Read with EOF")
+	return n, io.EOF
+}
+
+type ConnectionGroup struct {
+	messages       chan *RequestMessage
+	responseMap    map[uint32]chan *ResponseMessage
```

Comment:
Such definition of `responseMap` is not going to evolve nicely. For now it suffices to have `chan *ResponseMessage` instead of `struct PendingResponse` or `struct Request`, but you are going to need a struct when you decide to implement request cancelation.

I think `map[uint32]*Request` (the name `struct Request` may be different) is preferable as you will be able to create an API similar to that of `http.Client`:
```
req := gcsClient.NewRequest(...)
resp := gcsClient.Do(req)
...
res, err := resp.Wait()
...
```